pkg/migration/manager.go

```diff
@@ -0,0 +1,283 @@
+// the cluster.
+func (h *Helper) EveryNode(ctx context.Context, op serverpb.EveryNodeOp) error {
+	nodeIDs, err := h.RequiredNodes(ctx)
+	if err != nil {
+		return err
+	}
+
+	// TODO(irfansharif): We can/should send out these RPCs in parallel.
+	log.Infof(ctx, "executing op=%s on nodes=%s", redact.Safe(op.Op()), redact.Safe(nodeIDs))
+	for _, nodeID := range nodeIDs {
+		conn, err := h.dialer.Dial(ctx, nodeID, rpc.DefaultClass)
+		if err != nil {
+			return err
+		}
+
+		req := &serverpb.EveryNodeRequest{}
+		req.Request.SetInner(op)
+		client := serverpb.NewAdminClient(conn)
+		if _, err := client.EveryNode(ctx, req); err != nil {
+			return err
+		}
+	}
+
+	// TODO(irfansharif): We'll need to check RequiredNodes again to make sure
+	// that no new nodes were added to the cluster in the interim, and loop back
+	// around if so.
+	return nil
+}
+
+// MigrateTo runs the set of migrations required to upgrade the cluster version
+// to the provided target version.
+func (m *Manager) MigrateTo(ctx context.Context, targetV roachpb.Version) error {
+	// TODO(irfansharif): Should we inject every ctx here with specific labels
+	// for each migration, so they log distinctly? Do we need an AmbientContext?
+	ctx = logtags.AddTag(ctx, "migration-mgr", redact.Safe(targetV))
+
+	// TODO(irfansharif): We'll need to acquire a lease here and refresh it
+	// throughout during the migration to ensure mutual exclusion.
+
+	// TODO(irfansharif): We'll want to either write to a KV key to record the
+	// version up until which we've already migrated to, or consult the system
+	// table mentioned above. Perhaps it makes sense to consult any given
+	// `StoreClusterVersionKey`, since the manager here will want to push out
+	// cluster version bumps for vX before attempting to migrate into vX+1.
+
+	// TODO(irfansharif): After determining the last completed migration, if
+	// any, we'll be want to assemble the list of remaining migrations to step
+	// through to get to targetV. For now we've hard-coded this list.
+	var vs []roachpb.Version
+
+	// Hacks to introduce fake migrations on the fly.
+	//
+	// TODO(irfansharif): Remove these, and remove purge VersionNoopMigration
+	// from history.
+	Registry[targetV] = GenerateFakeMigrationFor(targetV)
+	vs = append(vs, targetV)
+
+	for _, version := range vs {
+		h := &Helper{Manager: m}
+		// Push out the version gate to every node in the cluster. Each node
+		// will persist the version, bump the local version gates, and then
+		// return. The migration associated with the specific version can assume
+		// that every node in the cluster has the corresponding version
+		// activated.
+		{
+			// First sanity check that we'll actually be able to perform the
+			// cluster version bump, cluster-wide.
+			//
+			// TODO(irfansharif): Do we want to worry about the case where we
+			// validate N nodes running the right binary versions, and the user
+			// adding another node that would no longer be able to? As written,
```

Comment:
Don't quite understand, you mean the case where a node joins after the validation but before the bump?

I thought we discussed this during the RFC. We need to roll out the cluster version (at least) twice. When we've done it once, we know that all the nodes we saw the first time around have the updated version, so they will only let new nodes join at that version. But some could've snuck in, but that's ok - we get the list of nodes again, do it one more time, and see if we caught any new nodes. If we didn't, done. If we did, do it again until it stabilizes (as that new node might've let yet another new node slip in).

Reply:
> you mean the case where a node joins after the validation but before the bump?

Yes, but I wasn't even thinking about the safety of the migration, I'm thinking about the safety of the version bump itself. Specifically the hazard I'm worried about is the following:
- N nodes running v21.1
- N nodes rolled into v21.2 binaries, but with active cluster version still as v21.1
- Validation for setting active cluster version to v21.2 checks to see that all nodes are running v21.2 binaries
- A new node is added to the cluster, but running binary v21.1
- We start bumping the cluster gates to v21.2, despite there now being a v21.1 node in the cluster.

To be clear, this isn't something we protect against today. Even with this dedicated ValidateTargetClusterVersion rpc, we'll not be fully protecting against it, it'll just be best effort. So perhaps that's ok.

Reply:
Added fence versions in #57186, which handles the hazards above.